### DISCUSSION TITLE
How to implement syntax highlighting?

### DISCUSSION THREAD

**Discussion by Charliego3** [krzyzanowskim/STTextView#6] [Questions]
like markdown, go ...

**Comment by vgorloff** [comment]
Ideally, if language developers will provide implementation of the Language Server Protocol (https://microsoft.github.io/language-server-protocol/). Then you can hook up data which comes from LSP into TextView and apply appropriate rendering attributes.

**Comment by krzyzanowskim** [comment]
With a help of https://github.com/ChimeHQ/Neon with tree-sitter parser and grammars is one way. I plan to add tree-sitter support  to the STTextView at one point.

**Comment by iliasaz** [comment]
Would love to see tree-sitter or Neon integration with STTextView, especially for large docs / low latency use case.

**Reply by krzyzanowskim** [reply]
me too. I don't think @mattmassicotte himself has time to contribute here. I didn't find time myself to sit on it yet, but that's something I would like to add as a plugin for the text view.

**Comment by mattmassicotte** [comment]
First, just a quick clarification. The semantic tokens feature of Language Server Protocol can be used to **augment** existing highlighting, but it is not a great system on its own. Very high latency, often does not support incremental edits, and typically has a relatively small file size limitation. Think of it as a nice secondary system.

Now, I made Neon because I've come to believe that it doesn't make sense for syntax highlighting to be within a view. It is much more flexible to keep the view and the data the view renders (the highlighted text) separate. Neon can probably just plug right into STTextView. And if not, I would be very happy to help make that work.

At a higher level, I'd love to see STTextView be a very good text view, and to focus on that. I think that highlighting can be external, and both it and the view will end up better for it.

**Comment by krzyzanowskim** [comment]
Use the dedicated plugin [STTextView-Plugin-Neon](https://github.com/krzyzanowskim/STTextView-Plugin-Neon) it works great!

**Reply by Qubik65536** [reply]
I found out that most of the plugins are macOS exclusive, is there a plan of supporting iOS/iPadOS? If I want to port it myself, where should I get started?

**Reply by krzyzanowskim** [reply]
That would be great to have help to port that to iOS. It should be feasible as the main dependency https://github.com/ChimeHQ/Neon/blob/main/Package.swift support iOS. The main challenge may be iOS plugin API as it is today, but that can be easily enchanced to add missing pieces. I'd say go for it!

**Reply by Qubik65536** [reply]
> That would be great to have help to port that to iOS. It should be feasible as the main dependency [ChimeHQ/Neon@`main`/Package.swift](https://github.com/ChimeHQ/Neon/blob/main/Package.swift) support iOS. The main challenge may be iOS plugin API as it is today, but that can be easily enchanced to add missing pieces. I'd say go for it!

@krzyzanowskim I’m glad that I can help with the port, however I’m currently not very familiar with these api, and need some help with it. Where should I get started to enhance the iOS plugin API?

**Reply by krzyzanowskim** [reply]
there's ni iOS plugin as of yet, you're the first. Steps:
1. Chekout that macOS template https://github.com/krzyzanowskim/STTextView-Plugin-Template
2. Change platform to iOS in package https://github.com/krzyzanowskim/STTextView-Plugin-Template/blob/main/Package.swift
3. Try to build DummyPlugin by commenting out what doesn't work https://github.com/krzyzanowskim/STTextView-Plugin-Template/blob/main/Sources/DummyPlugin/DummyPlugin.swift

Now you have a buildable foundation

4. Checkout Neon plugin https://github.com/krzyzanowskim/STTextView-Plugin-Neon
5. Look at it and port/copy code to iOS

Ideally both platforms can be supported from the same package using build dependencies 
![Screenshot 2024-09-02 at 19 55 09](https://github.com/user-attachments/assets/d16cd03c-5528-4586-a202-1949a34093b7)

for reference you can use the way I choosed when build multiplatform STTextView package https://github.com/krzyzanowskim/STTextView/blob/main/Package.swift

**Reply by Qubik65536** [reply]
> there's ni iOS plugin as of yet, you're the first. Steps:
> 
> 1. Chekout that macOS template [krzyzanowskim/STTextView-Plugin-Template](https://github.com/krzyzanowskim/STTextView-Plugin-Template)
> 2. Change platform to iOS in package [krzyzanowskim/STTextView-Plugin-Template@`main`/Package.swift](https://github.com/krzyzanowskim/STTextView-Plugin-Template/blob/main/Package.swift)
> 3. Try to build DummyPlugin by commenting out what doesn't work [krzyzanowskim/STTextView-Plugin-Template@`main`/Sources/DummyPlugin/DummyPlugin.swift](https://github.com/krzyzanowskim/STTextView-Plugin-Template/blob/main/Sources/DummyPlugin/DummyPlugin.swift)
> 
> Now you have a buildable foundation
> 
> 4. Checkout Neon plugin [krzyzanowskim/STTextView-Plugin-Neon](https://github.com/krzyzanowskim/STTextView-Plugin-Neon)
> 5. Look at it and port/copy code to iOS
> 
> Ideally both platforms can be supported from the same package using build dependencies ![Screenshot 2024-09-02 at 19 55 09](https://private-user-images.githubusercontent.com/758033/363766314-d16cd03c-5528-4586-a202-1949a34093b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzMDA0NTUsIm5iZiI6MTcyNTMwMDE1NSwicGF0aCI6Ii83NTgwMzMvMzYzNzY2MzE0LWQxNmNkMDNjLTU1MjgtNDU4Ni1hMjAyLTE5NDlhMzQwOTNiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwMlQxODAyMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTQzZGYwZDFlMDY4NjMwOGY2NmYxMzc1N2I3YmFiZGM0NTc3MmU1MTZmY2VmZmRhNDdjNTRhOTk4NWU5YWQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.mMk0fozfwnNRnGo0xcVcAPXgjTcqumRJhCOg5ksIQHw)
> 
> for reference you can use the way I choosed when build multiplatform STTextView package [`main`/Package.swift](https://github.com/krzyzanowskim/STTextView/blob/main/Package.swift)

Got it, I'll check it, thanks!

**Reply by Qubik65536** [reply]
> there's ni iOS plugin as of yet, you're the first. Steps:
> 
> 1. Chekout that macOS template [krzyzanowskim/STTextView-Plugin-Template](https://github.com/krzyzanowskim/STTextView-Plugin-Template)
> 2. Change platform to iOS in package [krzyzanowskim/STTextView-Plugin-Template@`main`/Package.swift](https://github.com/krzyzanowskim/STTextView-Plugin-Template/blob/main/Package.swift)
> 3. Try to build DummyPlugin by commenting out what doesn't work [krzyzanowskim/STTextView-Plugin-Template@`main`/Sources/DummyPlugin/DummyPlugin.swift](https://github.com/krzyzanowskim/STTextView-Plugin-Template/blob/main/Sources/DummyPlugin/DummyPlugin.swift)
> 
> Now you have a buildable foundation
> 
> 4. Checkout Neon plugin [krzyzanowskim/STTextView-Plugin-Neon](https://github.com/krzyzanowskim/STTextView-Plugin-Neon)
> 5. Look at it and port/copy code to iOS
> 
> Ideally both platforms can be supported from the same package using build dependencies ![Screenshot 2024-09-02 at 19 55 09](https://private-user-images.githubusercontent.com/758033/363766314-d16cd03c-5528-4586-a202-1949a34093b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzMDc5NDQsIm5iZiI6MTcyNTMwNzY0NCwicGF0aCI6Ii83NTgwMzMvMzYzNzY2MzE0LWQxNmNkMDNjLTU1MjgtNDU4Ni1hMjAyLTE5NDlhMzQwOTNiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwMlQyMDA3MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hM2VjYmE1NzIxZmU5YTI4MDI1YWI1MjEzMzllNDUyZTI4OTY2NzUzN2IwYWU0MzExMWMyY2E5M2E0ZmEzMjE0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-GE0NuvUeGQ70FsvbZzaQyJ9HVKGGDzDh9eueLjO4t0)
> 
> for reference you can use the way I choosed when build multiplatform STTextView package [`main`/Package.swift](https://github.com/krzyzanowskim/STTextView/blob/main/Package.swift)

I'm trying to learn how this works, now I got the dummy plugin compiled and even running with the SwiftUI demo in this repo with

```swift
TextView(
	text: $text,
	selection: $selection,
	options: [.wrapLines, .highlightSelectedLine, .showLineNumbers],
	plugins: [DummyPlugin()]
)
```

However, as I uncommented some of the print statement (e.g. the one in `didChangeText`), I still can't see if it is in effect because there are nothing printed out. Is that normal? The dummy plugin now is basically just:

```swift
import Foundation
import UIKit

import STTextView

public struct DummyPlugin: STPlugin {

    public init() { }

    public func setUp(context: any Context) {
        context.events.onWillChangeText(willChangeText)
        context.events.onDidChangeText(didChangeText)
        context.events.shouldChangeText(shouldChangeText)
        context.events.onContextMenu(contextMenu)
    }

    private func willChangeText(in affectedRange: NSTextRange) {
         print("will change handler!")
    }

    private func didChangeText(in affectedRange: NSTextRange, replacementString: String?) {
         print("did change handler from plugin!")
    }

    private func shouldChangeText(in textRange: NSTextRange, replacementString: String?) -> Bool {
        // if replacementString == "a" {
        //    return false
        // } 
        return true
    }

    private func contextMenu(_ location: NSTextLocation, _ contentManager: NSTextContentManager) -> UIMenu {
        let menu = UIMenu(title: "Dummy Plugin", children: [])
//        menu.autoenablesItems = false
//        menu.addItem(withTitle: "Dummy Action", action: nil, keyEquivalent: "")
//        menu.addItem(withTitle: "Smart Action", action: nil, keyEquivalent: "")
        return menu
    }
}
```

**Reply by krzyzanowskim** [reply]
> Is that normal?

that's not normal. maybe plugins support is broken on iOS. I'll take a look at it tomorrow, unless you find the cause earlier

**Reply by Qubik65536** [reply]
> > Is that normal?
> 
> that's not normal. maybe plugins support is broken on iOS. I'll take a look at it tomorrow, unless you find the cause earlier

Ok, thanks! I'll also try to take a look at it.

**Reply by Qubik65536** [reply]
> > Is that normal?
> 
> that's not normal. maybe plugins support is broken on iOS. I'll take a look at it tomorrow, unless you find the cause earlier

Is there any updates? I couldn't figure out what's wrong :(

**Reply by Qubik65536** [reply]
I tried to dig a little bit dipper, and it seems that `setUp(context: any Context)` of the plugin has never been executed. And it seems that the UIKit SwiftUI version of TextView had the plugin loading disabled?

https://github.com/krzyzanowskim/STTextView/blob/8d5d37665d8055b200a83e733806914029ab8401/Sources/STTextViewSwiftUIUIKit/TextView.swift#L93-L95

/cc @krzyzanowskim
